src/hooks/useSearchParamState.ts

```diff
@@ -105,9 +106,40 @@ const getAllCurrentParams = (searchParams: URLSearchParams) => {
   return allUrlParams;
 };
 
-const useSearchParamsState = (paramsDefinition: SearchParamsStateType) => {
+/**
+ * Works like useState but stores the state in URL search params. Adapted
+ * from https://github.com/jschwindt/react-use-search-params-state/tree/main
+ *
+ * @param paramsDefinition maps search param names to their definitions,
+ * indicating their type and default value
+ * @param initial initial values to set in the search params
+ */
+const useSearchParamsState = (
+  paramsDefinition: SearchParamsStateType,
+  initial?: URLSearchParamsInit
+) => {
+  // `initial` is NOT passed to `useSearchParams` here; useSearchParams's defaultInit prop
+  // auto-populates its internal state, but not the actual search params that appear in the url.
   const [searchParams, setSearchParams] = useSearchParams();
 
+  // Auto-populate the search params with the initial value
+  useEffect(() => {
+    if (!initial) return;
+
+    const toUpdate: Record<string, any> = Object.fromEntries(
+      // Find any key/value pairs that are in initial, but not in searchParams.
+      Object.entries(initial).filter(([key]) => !searchParams.has(key))
+    );
+
+    if (Object.keys(toUpdate).length) {
+      // This unfortunately needs to cause an extra render when setting searchParams to the initial value
+      setSearchParams(toUpdate, { replace: true });
+    }
+
+    // Only run on mount, so that we don't prevent the user from overwriting the initial values.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
   const values = useMemo(
```

Comment:
Poking around at https://github.com/greenriver/hmis-frontend/pull/1076/files#r1982130317 , `values` here should ideally represent the "initial" value when this hook first renders. I didn't road-test this much, but here's how I was proof-of-concepting it (`mounted` as a boolean piece of state that got set in the useEffect)

```js
  const values = useMemo(() => {
    if (mounted) {
      return getValues(paramsDefinition, searchParams);
    }
    if (initial) {
      // if the above useEffect hasn't run yet, set the initial values right away
      return getValues(paramsDefinition, createSearchParams(initial));
    }
    return {};
  }, [mounted, paramsDefinition, searchParams, initial]);

```

Reply:
Thanks, cool! This seems to work well, I made a small modification to avoid a warning about "changing an uncontrolled input to be controlled"